### PR TITLE
Audio PR - Augmentation support [ Spectrogram ]

### PR DESCRIPTION
**Pull request by swetha097**
- Add Spectrogram Filter support in rocAL

MIVisionX PR Link : [PR#1319](https://github.com/ROCm/MIVisionX/pull/1319)
NOTE: [PR#123](https://github.com/ROCm/rocAL/pull/123) to be merged before this PR

### CODE REVIEW

**Comment by rrawther** [comment]
@swetha097 please add the MIVisionX version check and disable this feature until the next version of MIVisionX where audio feature will be enabled

File: tests/cpp_api/audio_tests/audio_tests.cpp [hunk 1]
```diff
@@ -45,35 +45,47 @@ bool verify_output(float *dst_ptr, long int frames, std::string case_name) {
...
+            out_val = dst_ptr[i * max_channels + j];
+            bool invalid_comparison = ((out_val == 0.0f) && (ref_val != 0.0f));
+            if (!invalid_comparison && std::abs(out_val - ref_val) < 1e-20)
+                matched_indices += 1;
+        }
     }
 
-    std::cout << std::endl << "Results for Test case: " << std::endl;
-    if ((matched_indices == frames) && matched_indices != 0) {
+    std::cout << std::endl
```
Comment (rrawther) [inline tests/cpp_api/audio_tests/audio_tests.cpp:86]: please remove this emply line. The current way is correct

File: tests/cpp_api/audio_tests/audio_tests.cpp [hunk 2]
```diff
@@ -132,12 +144,17 @@ int test(int test_case, const char *path, int qa_mode, int downmix, int gpu) {
         return -1;
     }
 
-    std::cout << "Running LABEL READER" << std::endl;
-    rocalCreateLabelReader(handle, path);
+    std::string file_list_path = "";  // User can modify this with the file list path if required
```
Comment (rrawther) [inline tests/cpp_api/audio_tests/audio_tests.cpp:147]: std::string constructor will create empty string by default. So = "" is not required. Please remove all such cases in other instances

File: tests/python_api/audio_unit_test.py [hunk 3]
```diff
@@ -116,8 +155,9 @@ def main():
     if not rocal_cpu:
         print("The GPU support for Audio is not given yet. Running on CPU")
         rocal_cpu = True
-    if audio_path == "":
-        audio_path = f'{rocal_data_path}/rocal_data/audio/wav/'
+    if audio_path == "" and file_list == "":
```
Comment (rrawther) [inline tests/python_api/audio_unit_test.py:158]: empty() is better here
```suggestion
    if audio_path == "" and file_list == "":
```

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 please resolve merge conflicts

File: rocAL/include/api/rocal_api_augmentation.h [hunk 4]
```diff
@@ -1098,4 +1098,50 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSSDRandomCrop(RocalContext context, R
...
+ * \ingroup group_rocal_augmentations
+ * \param [in] context Rocal context
+ * \param [in] input Input Rocal tensor
+ * \param [in] is_output Sets to True if the output tensor is part of the graph output
+ * \param [in] preemph_coeff Preemphasis coefficient
+ * \param [in] preemph_border_type Border value policy. Possible values are "zero", "clamp", "reflect".
+ * \param [in] output_datatype The data type of the output tensor
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalPreEmphasisFilter(RocalContext context,
```
Comment (LakshmiKumar23) [inline rocAL/include/api/rocal_api_augmentation.h:1111]: Make sure to add ROCAL_AUDIO flag wherever required in this PR and others

File: rocAL/include/api/rocal_api_meta_data.h [hunk 5]
```diff
@@ -36,9 +36,10 @@ THE SOFTWARE.
  * \ingroup group_rocal_meta_data
  * \param [in] rocal_context rocal context
  * \param [in] source_path path to the folder that contains the dataset or metadata file
+ * \param file_list_path is the path to file list that contains the file names and its corresponding labels
  * \return RocalMetaData object, can be used to inquire about the rocal's output (processed) tensors
  */
-extern "C" RocalMetaData ROCAL_API_CALL rocalCreateLabelReader(RocalContext rocal_context, const char* source_path);
+extern "C" RocalMetaData ROCAL_API_CALL rocalCreateLabelReader(RocalContext rocal_context, const char* source_path, const char* file_list_path = "");
```
Comment (LakshmiKumar23) [inline rocAL/include/api/rocal_api_meta_data.h:42]: we should be consistent with initializing source_path and file_list_path

File: rocAL/include/augmentations/audio_augmentations/node_spectrogram.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#pragma once
+#include "pipeline/graph.h"
+#include "pipeline/node.h"
+#include "rocal_api_types.h"
+
+/// @brief Generates hann window for spectrogram
+/// @param output 
+/// @param window_size 
+inline void hann_window(float *output, int window_size) {
+    double a = (2.0 * M_PI) / window_size;
```
Comment (LakshmiKumar23) [inline rocAL/include/augmentations/audio_augmentations/node_spectrogram.h:32]: check for divide by 0

**Review by rrawther — CHANGES_REQUESTED** [review]
Added review comments

File: tests/cpp_api/audio_tests/audio_tests.cpp [hunk 7]
```diff
@@ -132,12 +144,17 @@ int test(int test_case, const char *path, int qa_mode, int downmix, int gpu) {
...
+    if (qa_mode) {                    // setting the default file list path from ROCAL_DATA_PATH
+        file_list_path = std::string(std::getenv("ROCAL_DATA_PATH")) + "rocal_data/audio/wav_file_list.txt";
+    }
+
+    std::cout << ">>>>>>> Running LABEL READER" << std::endl;
```
Comment (LakshmiKumar23) [inline tests/cpp_api/audio_tests/audio_tests.cpp:152]: please remove the extra ">>>>" they are not needed

File: tests/cpp_api/audio_tests/audio_tests.cpp [hunk 8]
```diff
@@ -149,6 +166,22 @@ int test(int test_case, const char *path, int qa_mode, int downmix, int gpu) {
...
+            std::cout << "Running PREEMPHASIS" << std::endl;
+            case_name = "preemphasis_filter";
+            RocalTensorOutputType tensorOutputType = RocalTensorOutputType::ROCAL_FP32;
+            RocalAudioBorderType preemph_border_type = RocalAudioBorderType::ROCAL_CLAMP;
+            RocalFloatParam p_preemph_coeff = rocalCreateFloatParameter(0.97);
+            rocalPreEmphasisFilter(handle, decoded_output, true, p_preemph_coeff, preemph_border_type, tensorOutputType);
+
+        } break;
+        case 2: {
+            std::cout << ">>>>>>> Running SPECTROGRAM" << std::endl;
```
Comment (LakshmiKumar23) [inline tests/cpp_api/audio_tests/audio_tests.cpp:179]: same as above, please refer to other tests

File: tests/python_api/README.md [hunk 9]
```diff
@@ -141,3 +141,4 @@ python3 audio_unit_test.py --audio_path=<path_to_data> --test_case <case(0-1)> -
 **Available Test Cases**
```
Comment (LakshmiKumar23) [inline tests/python_api/README.md:141]: Update line 138 to reflect `case 0-2`

**Comment by LakshmiKumar23** [comment]
@fiona-gladwin added a minor comment. Can be merged after that. Tests are passing

File: tests/python_api/audio_unit_test.py (hunk 3 above)
Comment (fiona-gladwin) [reply tests/python_api/audio_unit_test.py]: Changed to not

File: tests/python_api/README.md (hunk 9 above)
Comment (fiona-gladwin) [reply tests/python_api/README.md:141]: Done

**Review by LakshmiKumar23 — APPROVED** [review]
Spectrogram tests pass

**Review by rrawther — APPROVED** [review]

**Pull request merged**